sources/parsers/parsers.ml

```diff
@@ -47,37 +47,63 @@ let register_parser ~lang new_parser =
     end;
   parsers := (lang, new_parser) :: !parsers
 
+let debug_no_parser_for_no_extension () =
+  if verbose() then
+    eprintf
+      "error: no extension found, can't infer input format@."
+
 let debug_no_parser_for_extension s =
   if verbose() then
     eprintf
-      "error: no parser registered for ext %S. Using default lang ...@." s
+      "error: no parser registered for ext %S @." s
 
 let debug_no_parser_for_input_format s =
   if verbose() then
     eprintf
       "error: no parser registered for the provided input format %S ?@." s
 
+let get_input_parser () =
+  match Options.input_format () with
+  | Options.Native -> begin
+      try List.assoc ".ae" !parsers
+      with Not_found ->
+        debug_no_parser_for_input_format "ae";
+        exit 1
+    end
+  | Options.Smtlib2 -> begin
+      try List.assoc ".smt2" !parsers
+      with Not_found ->
+        debug_no_parser_for_input_format "smtlib";
+        exit 1
+    end
+  | Options.Why3 -> begin
+      try List.assoc ".why3" !parsers
+      with Not_found ->
+        debug_no_parser_for_input_format "why3";
+        exit 1
+    end
+  | Options.Unknown s -> begin
+      debug_no_parser_for_input_format s;
+      exit 1
+    end
+
 let get_parser lang_opt =
-  let i = Options.input_format () in
   match lang_opt with
-  | Some lang ->
-    begin
-      try List.assoc lang !parsers
-      with Not_found ->
-        debug_no_parser_for_extension lang;
-        try List.assoc i !parsers
+  | Some lang -> begin
+      if Options.infer_input_format () then
```

Comment:
You should add a warning here I think, or at least warn when a user has specified both to infer the file format and a specific file format.

Reply:
I you had read carefully this PR you should know that input format inference is only done when no input format is set.

Reply:
The problem overall, is that there are way too much places when the inference flag is used to change behaviour, optimally there should only be one place where it is used.
Here, there seems to be a language provided so why is there a need to have different behaviour based on whether inference is activated ?

Reply:
Too much places, how's that ?

Reply:
There are at least three invocations of `Options.infer_input_format` in this file (two in this function, and one later): basically, I'm saying this is two more than should be. We need one such call at the place where we decide which parser should be used, but I don't see why we would need more than that one and only call. And neither the code nor the (lack of) documentation around explains these extra calls.